website/blog/2019-02-22-shadow-dom-support.md

```diff
@@ -0,0 +1,186 @@
+  }
+
+  get host() {
+    // Beginning with the browser object, reselect each element
+    return this.selectors_.reduce((element, selector) => element.$(selector), browser);
+  }
+}
+
+module.exports = Component;
+```
+
+We can then write a subclass for our app-login component:
+
+```javascript
+const Component = require('./component');
+
+class Login extends Component {
+
+  get usernameInput() {
+    return this.host.shadow$('input #username');
+  }
+
+  get passwordInput() {
+    return this.host.shadow$('input[type=password]');
+  }
+
+  get submitButton() {
+    return this.login.shadow$('button[type=submit]');
+  }
+
+  login(username, password) {
+    this.usernameInput.setValue(username);
+    this.passwordInput.setValue(password);
+    this.submitButton.click();
+  }
+}
+
+module.exports = Login;
+```
+
+Finally, we can use the component object inside our login page object:
+
+```javascript
+const Login = require('./components/login');
+
+class LoginPage {
+
+  open() {
+    browser.url('/login');
+  }
+
+  get app() {
+    return browser.$('my-app');
+  }
+
+  get loginComponent() {
+    // return a new instance of our login component object
+    return new Login(this.app.$('app-login'));
+  }
+
+}
+```
+This component object can now be used in tests for any page or section of your app that uses an app-login web component, without having to know about how that component is structured. If you later decide to change the internal structure of the web component, you only need to update the component object.
+
+
+## Future
+Currently Selenium does not provide native support for shadow DOM, and the w3c webdriver spec writers are still up in the air as to how it should be done. Once the spec is finalized, WebdriverIO will implement the spec. There's a decent chance that the `shadow` commands will change under the hood, but I'm pretty confident that they're usage will be the same as it is today, and that test code that uses them will need little to no refactoring.
+
+## Browser Support
+IE11-Edge: Shadow DOM is not supported in IE or Edge, but can be polyfilled. The shadow commands work great with the polyfills.
+
+Firefox: Calling `setValue(value)` on an input field in Firefox results in an error, complaining that the input is "not reachable by keyboard". A workaround for now is to use a custom command (or method on your component object) that sets the input field's value via `browser.execute(function)`.
+
+Safari: WebdriverIO has some safety mechanisms to help mitigate issues with stale element references. This is a really nice feature but unfortunately Safari's webdriver does not provide the proper error response when attempting to interact with what in other browsers, is a stale element reference. This is unfortunate but at the same time, it's generally a bad practice to cache element references. Stale element references are typically completely mitigated by using the page and component object patterns outlined above.
```

Comment:
Can you create an issue for this with some logs about the error message? It should be easy to detect safaridrivers stale element error and handle it similar to Chrome or others.